### ISSUE TITLE
ADIOS2 v2.7.1: Spack Build on Summit Breaks IB

### ISSUE THREAD

**Issue by ax3l** [ornladios/ADIOS2#2836]
Hi,

I am using the E4S Spack package of ADIOS2 v2.7.1 on Summit (`/sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/`). We are really happy about that opportunity, because it saves us from having to maintain and compile ADIOS2 ourselves for the whole WarpX user base, which eases our support burden a lot and gives us a fast-track to new ADIOS2 releases for production runs.

Unfortunately, I think that some components of ADIOS2 forget to declare infiniband (IB) / libfabric or MPI-related dependencies, because the moment I link ADIOS2 to my application (ECP WarpX), I get the following errors on startup of our app, even if I don't use any ADIOS functionality:
```
Number of active IB device ports not supported
[g12n09:264854] Error: common_pami.c:1094 - ompi_common_pami_init() 0: Unable to create 1 PAMI communication context(s) rc=1
--------------------------------------------------------------------------
No components were able to be opened in the pml framework.

This typically means that either no components of this type were
installed, or none of the installed components can be loaded.
Sometimes this means that shared libraries required by these
components are unable to be found/loaded.

  Host:      g12n09
  Framework: pml
--------------------------------------------------------------------------
[g12n09:264854] PML pami cannot be selected
```

I will also open an OLCF ticket for this and link to here for coordination.

In parallel, it would be perfect if you @pnorbert @chuckatkins @eisenhauer @williamfgc @sklasky et al. can coordinate with OLCF support, your Spack package and E4S to make sure the SST component's CMake logic in combination with the `depends_on` declarations in your Spack package declare the right dependencies.

Furthermore, the Spack ADIOS2 package could declare simple acceptance tests (built-time checks and smoke-tests) to make sure system people can automatically validate your compiled package really works. Those tests can be a small sub-set of your regular `ctests` to check MPI et al. truly works.

### References
- [Spack's ADIOS2 package.py](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/adios2/package.py)
- Two examples to add check/smoke-tests: [openPMD-api](https://github.com/spack/spack/blob/37a1885deb8142eaa8ec9bf5e2f18d14e81a1729/var/spack/repos/builtin/packages/openpmd-api/package.py#L134-L152) or [WarpX](https://github.com/spack/spack/blob/37a1885deb8142eaa8ec9bf5e2f18d14e81a1729/var/spack/repos/builtin/packages/warpx/package.py#L139-L187)

### ldd libadios2_cxx11_mpi.so

<details>
$ ldd $OLCF_ADIOS2_ROOT/lib64/libadios2_cxx11_mpi.so                                
	linux-vdso64.so.1 (0x0000200000050000)
	libadios2_cxx11.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/libadios2_cxx11.so.2 (0x00002000000a0000)
	libadios2_core_mpi.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/libadios2_core_mpi.so.2 (0x0000200000260000)
	libadios2_core.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/libadios2_core.so.2 (0x00002000003f0000)
	libpthread.so.0 => /lib64/power9/libpthread.so.0 (0x0000200000b20000)
	libmpiprofilesupport.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpiprofilesupport.so.3 (0x0000200000b70000)
	libmpi_ibm.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpi_ibm.so.3 (0x0000200000ba0000)
	libstdc++.so.6 => /autofs/nccs-svm1_sw/summit/gcc/9.3.0-2/lib64/libstdc++.so.6 (0x0000200000d60000)
	libm.so.6 => /lib64/power9/libm.so.6 (0x0000200000ff0000)
	libgcc_s.so.1 => /autofs/nccs-svm1_sw/summit/gcc/9.3.0-2/lib64/libgcc_s.so.1 (0x0000200001120000)
	libc.so.6 => /lib64/power9/libc.so.6 (0x0000200001160000)
	libadios2_taustubs.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_taustubs.so (0x0000200001370000)
	libblosc.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/c-blosc-1.21.0-s6xjxljpolgxpoh73tuwgjhvtcy2fgh4/lib/libblosc.so.1 (0x00002000013a0000)
	libbz2.so.1.0 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/bzip2-1.0.8-rgx7teyivl2lbzzznj3vyy5j64mw3n7g/lib/libbz2.so.1.0 (0x00002000013d0000)
	libzfp.so.0 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zfp-0.5.5-va2z6zttx57egp3n4c4sddhx2mp6ysc4/lib64/libzfp.so.0 (0x0000200001410000)
	libSZ.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/sz-2.1.11.1-7hgfrsk5r7e7yoyupgiiaprbh763igee/lib64/libSZ.so (0x0000200001470000)
	libpng16.so.16 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libpng-1.6.37-q2i4njub2bbhfheqpqjcwjbta6az5ret/lib/libpng16.so.16 (0x0000200001580000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002000015f0000)
	libfabric.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libfabric-1.12.1-fcjy5sxzwx2fiflbpirvo5qzwd73pa5p/lib/libfabric.so.1 (0x0000200001620000)
	libadios2_evpath.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_evpath.so (0x0000200001730000)
	libadios2_ffs.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_ffs.so.1 (0x00002000017e0000)
	libadios2_atl.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_atl.so.2 (0x0000200001880000)
	libz.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zlib-1.2.11-5ylwliurlfdizyqq2h2ujwdvl44ukaef/lib/libz.so.1 (0x00002000018b0000)
	libzstd.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zstd-1.5.0-j6742s2uqme6gcdwuxhwbqajmbmuwpki/lib64/libzstd.so.1 (0x00002000018f0000)
	/lib64/ld64.so.2 (0x0000200000000000)
	librt.so.1 => /lib64/power9/librt.so.1 (0x00002000019f0000)
	libutil.so.1 => /lib64/libutil.so.1 (0x0000200001a20000)
	libhwloc_ompi.so.15 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libhwloc_ompi.so.15 (0x0000200001a50000)
	libevent_core-2.1.so.6 => /lib64/libevent_core-2.1.so.6 (0x0000200001ae0000)
	libevent_pthreads-2.1.so.6 => /lib64/libevent_pthreads-2.1.so.6 (0x0000200001b70000)
	libopen-rte.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-rte.so.3 (0x0000200001ba0000)
	libopen-pal.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-pal.so.3 (0x0000200001cc0000)
	liblz4.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/lz4-1.9.3-vfbfbfgzir4xbt3nlmbbkbl2rgyfkngq/lib/liblz4.so.1 (0x0000200001e20000)
	libgomp.so.1 => /autofs/nccs-svm1_sw/summit/gcc/9.3.0-2/lib64/libgomp.so.1 (0x0000200001e80000)
	librdmacm.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/rdma-core-34.0-grvfqvzut26nwzs5yhfku42hqeolx63k/lib64/librdmacm.so.1 (0x0000200001ef0000)
	libibverbs.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/rdma-core-34.0-grvfqvzut26nwzs5yhfku42hqeolx63k/lib64/libibverbs.so.1 (0x0000200001f30000)
	libatomic.so.1 => /sw/summit/gcc/9.3.0-2/lib/../lib64/libatomic.so.1 (0x0000200001f80000)
	libadios2_dill.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/../lib64/libadios2_dill.so.2 (0x0000200001fb0000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x0000200002010000)
	libnl-3.so.200 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libnl-3.3.0-2b2gen3crztivrqz4rnyafw5mpjib7cv/lib/libnl-3.so.200 (0x0000200002380000)
	libnl-route-3.so.200 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libnl-3.3.0-2b2gen3crztivrqz4rnyafw5mpjib7cv/lib/libnl-route-3.so.200 (0x00002000023d0000)
	libffi.so.7 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libffi-3.3-vgzin2dcgkwyvp5nq75ioktj7adpw35w/lib64/libffi.so.7 (0x0000200002480000)
</details>

**Comment by eisenhauer** [comment]
Well, the initial error, "Number of active IB device ports not supported" is unfamiliar to either me or google.  I guess I don't know whether or not to credit the fewer dependencies of warpx and libadios2_cxx11_mpi.so as the root cause or not.  Seems like those dependencies are usually transitive, but I don't know if spack does something funny there.  If I were guessing though I might guess that there might be a version conflict with whatever MPI relies upon for RDMA and whatever we're using.  @chuckatkins , maybe something like we resolved on rhea by loading the rdma-core module?

**Comment by ax3l** [comment]
I think that error originates from the bowels of SpectrumMPI, but likely triggered due to a dependency mismatch.

For what it's worth, this is the WarpX `ldd` list w/o ADIOS2 (21 direct loads):
```
	/lib64/ld64.so.2 (0x0000200000000000)
	libc.so.6 => /lib64/power9/libc.so.6 (0x0000200000cd0000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00002000012f0000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002000008a0000)
	libevent_core-2.1.so.6 => /lib64/libevent_core-2.1.so.6 (0x0000200000fd0000)
	libevent_pthreads-2.1.so.6 => /lib64/libevent_pthreads-2.1.so.6 (0x0000200001040000)
	libgcc_s.so.1 => /sw/summit/gcc/9.3.0-2/lib64/libgcc_s.so.1 (0x0000200000c90000)
	libgomp.so.1 => /sw/summit/gcc/9.3.0-2/lib64/libgomp.so.1 (0x0000200000070000)
	libhdf5.so.103 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/hdf5-1.10.7-4b77lkzgqo6zq6waojy3c24asrm5tu5h/lib/libhdf5.so.103 (0x0000200000340000)
	libhwloc_ompi.so.15 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libhwloc_ompi.so.15 (0x0000200000f40000)
	libm.so.6 => /lib64/power9/libm.so.6 (0x00002000008d0000)
	libmpi_ibm.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpi_ibm.so.3 (0x0000200000180000)
	libmpiprofilesupport.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpiprofilesupport.so.3 (0x0000200000150000)
	libopen-pal.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-pal.so.3 (0x0000200001190000)
	libopen-rte.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-rte.so.3 (0x0000200001070000)
	libpthread.so.0 => /lib64/power9/libpthread.so.0 (0x0000200000100000)
	librt.so.1 => /lib64/power9/librt.so.1 (0x0000200000ee0000)
	libstdc++.so.6 => /sw/summit/gcc/9.3.0-2/lib64/libstdc++.so.6 (0x0000200000a00000)
	libutil.so.1 => /lib64/libutil.so.1 (0x0000200000f10000)
	libz.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zlib-1.2.11-5ylwliurlfdizyqq2h2ujwdvl44ukaef/lib/libz.so.1 (0x0000200000860000)
	linux-vdso64.so.1 (0x0000200000050000)
```

and that's the list with ADIOS2 (44 direct loads):
```
	/lib64/ld64.so.2 (0x0000200000000000)
	libSZ.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/sz-2.1.11.1-7hgfrsk5r7e7yoyupgiiaprbh763igee/lib64/libSZ.so (0x0000200001ea0000)
	libadios2_atl.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_atl.so.2 (0x0000200002280000)
	libadios2_core.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/libadios2_core.so.2 (0x0000200001260000)
	libadios2_core_mpi.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/libadios2_core_mpi.so.2 (0x00002000010d0000)
	libadios2_cxx11.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/libadios2_cxx11.so.2 (0x0000200000840000)
	libadios2_cxx11_mpi.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/libadios2_cxx11_mpi.so.2 (0x0000200000810000)
	libadios2_dill.so.2 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/../lib64/libadios2_dill.so.2 (0x0000200002840000)
	libadios2_evpath.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_evpath.so (0x0000200002130000)
	libadios2_ffs.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_ffs.so.1 (0x00002000021e0000)
	libadios2_taustubs.so => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/adios2-2.7.1-vuplpxvgpbby2oe4sgpbtarlrg55ofxl/lib64/../lib64/../lib64/libadios2_taustubs.so (0x0000200001d80000)
	libatomic.so.1 => /sw/summit/gcc/9.3.0-2/lib/../lib64/libatomic.so.1 (0x0000200002810000)
	libblosc.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/c-blosc-1.21.0-s6xjxljpolgxpoh73tuwgjhvtcy2fgh4/lib/libblosc.so.1 (0x0000200001db0000)
	libbz2.so.1.0 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/bzip2-1.0.8-rgx7teyivl2lbzzznj3vyy5j64mw3n7g/lib/libbz2.so.1.0 (0x0000200001e00000)
	libc.so.6 => /lib64/power9/libc.so.6 (0x0000200000ec0000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00002000023b0000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002000006b0000)
	libevent_core-2.1.so.6 => /lib64/libevent_core-2.1.so.6 (0x0000200001a60000)
	libevent_pthreads-2.1.so.6 => /lib64/libevent_pthreads-2.1.so.6 (0x0000200001ad0000)
	libfabric.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libfabric-1.12.1-fcjy5sxzwx2fiflbpirvo5qzwd73pa5p/lib/libfabric.so.1 (0x0000200002020000)
	libffi.so.7 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libffi-3.3-vgzin2dcgkwyvp5nq75ioktj7adpw35w/lib64/libffi.so.7 (0x00002000029a0000)
	libgcc_s.so.1 => /sw/summit/gcc/9.3.0-2/lib64/libgcc_s.so.1 (0x0000200000e80000)
	libgomp.so.1 => /sw/summit/gcc/9.3.0-2/lib64/libgomp.so.1 (0x0000200000070000)
	libhdf5.so.103 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/hdf5-1.10.7-4b77lkzgqo6zq6waojy3c24asrm5tu5h/lib/libhdf5.so.103 (0x0000200000150000)
	libhwloc_ompi.so.15 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libhwloc_ompi.so.15 (0x00002000019d0000)
	libibverbs.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/rdma-core-34.0-grvfqvzut26nwzs5yhfku42hqeolx63k/lib64/libibverbs.so.1 (0x00002000027c0000)
	liblz4.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/lz4-1.9.3-vfbfbfgzir4xbt3nlmbbkbl2rgyfkngq/lib/liblz4.so.1 (0x0000200002720000)
	libm.so.6 => /lib64/power9/libm.so.6 (0x00002000006e0000)
	libmpi_ibm.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpi_ibm.so.3 (0x0000200000a30000)
	libmpiprofilesupport.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libmpiprofilesupport.so.3 (0x0000200000a00000)
	libnl-3.so.200 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libnl-3.3.0-2b2gen3crztivrqz4rnyafw5mpjib7cv/lib/libnl-3.so.200 (0x00002000028a0000)
	libnl-route-3.so.200 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libnl-3.3.0-2b2gen3crztivrqz4rnyafw5mpjib7cv/lib/libnl-route-3.so.200 (0x00002000028f0000)
	libopen-pal.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-pal.so.3 (0x0000200001c20000)
	libopen-rte.so.3 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/lib/libopen-rte.so.3 (0x0000200001b00000)
	libpng16.so.16 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/libpng-1.6.37-q2i4njub2bbhfheqpqjcwjbta6az5ret/lib/libpng16.so.16 (0x0000200001fb0000)
	libpthread.so.0 => /lib64/power9/libpthread.so.0 (0x0000200000100000)
	librdmacm.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/rdma-core-34.0-grvfqvzut26nwzs5yhfku42hqeolx63k/lib64/librdmacm.so.1 (0x0000200002780000)
	librt.so.1 => /lib64/power9/librt.so.1 (0x0000200001970000)
	libstdc++.so.6 => /sw/summit/gcc/9.3.0-2/lib64/libstdc++.so.6 (0x0000200000bf0000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00002000019a0000)
	libz.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zlib-1.2.11-5ylwliurlfdizyqq2h2ujwdvl44ukaef/lib/libz.so.1 (0x0000200000670000)
	libzfp.so.0 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zfp-0.5.5-va2z6zttx57egp3n4c4sddhx2mp6ysc4/lib64/libzfp.so.0 (0x0000200001e40000)
	libzstd.so.1 => /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/zstd-1.5.0-j6742s2uqme6gcdwuxhwbqajmbmuwpki/lib64/libzstd.so.1 (0x00002000022b0000)
	linux-vdso64.so.1 (0x0000200000050000)
```

**Comment by chuckatkins** [comment]
This is a known problem on summit and a mismatch between the rdma-core libraries used by spectrum-mpi and spack's libfabric.  I believe the admins built a specific module for us to fix this.  Try using the libfabric/1.7.0-sysrdma module instead of the default libfabric.

**Comment by ax3l** [comment]
But we can also try to reproduce this with https://github.com/ornladios/ADIOS2-Examples if you like.

In an interactive session via
```
cmake -S . -B build -DMPIEXEC_EXECUTABLE=$(which jsrun)
cmake --build build -j 64

cd build
ctest --output-on-failure
```

I can see the same issues.

**Comment by eisenhauer** [comment]
Just to clarify @ax3l, did you try ADIOS2-Examples with the libfabric/1.7.0-sysrdma as Chuck suggested and still see a failure?

**Comment by ax3l** [comment]
Not yet, I posted at the same time and then went in the weekend :)

**Comment by ax3l** [comment]
Thanks a lot for the quick suggestion @chuckatkins!
Just checked - it looks like the module does not exist anymore after the system upgrade to RHEL8.

If a custom Summit-specific libfabric is the solution: maybe the OLCF E4S team can register this version of libfabric as an [external spack package](https://spack.readthedocs.io/en/latest/build_settings.html#external-packages) for all their builds. That way, all packages built out use that specifically tweaked libfabric that you mentioned. (Found [this config](https://github.com/spack/spack-configs) but seems outdated.)

Just curious: is that an IBM or Mellanox fork of libfabric on Summit? Or were the changes/fixes upstreamed?

**Comment by chuckatkins** [comment]
> Thanks a lot for the quick suggestion @chuckatkins!
> Just checked - it looks like the module does not exist anymore after the system upgrade to RHEL8.

An easy workaround should be to just use `LD_PRELOAD`:
```bash
$ export LD_PRELOAD="/usr/lib64/libibverbs.so.1:/usr/lib64/librdmacm.so.1"
```

> If a custom Summit-specific libfabric is the solution
> ...
> Just curious: is that an IBM or Mellanox fork of libfabric on Summit? Or were the changes/fixes upstreamed?

The problem is a mismatch in the `rdma-core` libraries used by the IBM-provided Spectrum MPI RPM and the spack-provided libfabric.  Spectrum MPI is built against the system-provided rdma-core package while spack's libfabric is built against the spack-provided rdma-core package.  The result is that you end up with either MPI or libfabric working but not both in the same build.  Setting `LD_PRELOAD` to the system rdma-core libraries will ensure that those ones get used at runtime by spack's libfabric (the converse causes Spectrum MPI to segfault).

**Comment by mpbelhorn** [comment]
@chuckatkins @ax3l Following up on the OLCFHELP-3319 ticket, we put a module `libfabric/1.12.1-sysrdma` on summit yesterday afternoon that uses the system RDMA-core and have re-built our `adios2/2.7.1` modules against it. If you try it out, please update the OLCF ticket with any other issues you find.

**Comment by ax3l** [comment]
Just added to the OLCF ticket: when going to a few 100 nodes, we see new errors of the kind
```
 7: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libcollectives.so.3(_ZN4CCMI9Protocols7Barrier26MultiLeaderBarrierFactoryTINS1_19MultiLeaderBarrierTIN7LibColl10Interfaces15NativeInterfaceELNS4_15topologyIndex_tE0EEENS_17ConnectionManager13SimpleConnMgrEE8cb_asyncEPvSC_PKvjjmPPNS4_13PipeWorkQueueEPPFvSC_SC_16libcoll_result_tEPSC_+0x39c) [0x20000f3a0c7c]

 8: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libcollectives.so.3(_ZN7LibColl18NativeInterfaceP2PILb1ELb0EE20dispatch_mcast_shortEPvS2_PKvmS4_mjP11pami_recv_t+0x80) [0x20000f40d3d0]

 9: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libpami.so.3(_ZN4PAMI8Protocol4Send11EagerSimpleINS_6Device3IBV11PacketModelINS4_6DeviceELb1EEELNS1_15configuration_tE1EE15dispatch_packedEPvSA_mSA_SA_+0x4c) [0x20000ebf24ec]

10: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libpami.so.3(PAMI_Context_advancev+0x994) [0x20000ebea1b4]

11: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libcollectives.so.3(LIBCOLL_Advance_pami+0x34) [0x20000f33fb14]

12: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/pami_port/libcollectives.so.3(LIBCOLL_Advance+0x18) [0x20000f32a128]

13: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/spectrum_mpi/mca_coll_ibm.so(start_libcoll_blocking_collective+0x120) [0x20000f23fe50]

14: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/spectrum_mpi/mca_coll_ibm.so(mca_coll_ibm_barrier+0x70) [0x20000f2451d0]

15: /sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.3.0/spectrum-mpi-10.4.0.3-20210112-2s7kpbzydf6val7k2d3e6cz3zdhtcwlw/container/../lib/libmpi_ibm.so.3(MPI_Barrier+0x104) [0x20000a7d0b34]

16: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core_mpi.so.2(_ZNK6adios26helper11CommImplMPI7BarrierERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x24) [0x20000aea7c54]

17: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZNK6adios26helper4Comm7BarrierERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x2c) [0x20000b2382ac]

18: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios212transportman12TransportMan13MkDirsBarrierERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EERKS2_ISt3mapIS8_S8_St4lessIS8_ESaISt4pairIKS8_S8_EEESaISK_EEb+0x10c) [0x20000b4e137c]

19: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core6engine9BP4Writer14InitTransportsEv+0x25c) [0x20000b29f68c]

20: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core6engine9BP4Writer4InitEv+0x78) [0x20000b2a0248]

21: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core6engine9BP4WriterC2ERNS0_2IOERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_4ModeENS_6helper4CommE+0x308) [0x20000b2a0588]

22: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core2IO10MakeEngineINS0_6engine9BP4WriterEEESt10shared_ptrINS0_6EngineEERS1_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_4ModeENS_6helper4CommE+0xa0) [0x20000b191480]

23: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core_mpi.so.2(_ZNSt17_Function_handlerIFSt10shared_ptrIN6adios24core6EngineEERNS2_2IOERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS1_4ModeENS1_6helper4CommEEPSI_E9_M_invokeERKSt9_Any_dataS6_SE_OSF_OSH_+0x74) [0x20000aea56b4]

24: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core2IO4OpenERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_4ModeENS_6helper4CommE+0x540) [0x20000b18b5d0]

25: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/../lib64/libadios2_core.so.2(_ZN6adios24core2IO4OpenERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_4ModeE+0x78) [0x20000b18c168]

26: /sw/summit/ums/aph114/gcc-9.3.0-cxx14/spectrum-mpi-10.4.0.3-20210112/adios2-2.7.1-nosst/lib64/libadios2_cxx11.so.2(_ZN6adios22IO4OpenERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_4ModeE+0x168) [0x20000a610d38]
```

with the ADIOS module on Summit. The problem does not show up at smaller scale.

cc @mpbelhorn  @chuckatkins

**Comment by ax3l** [comment]
Opened a new ticket in OLCFHELP-3545 since the old one is closed.
I can also open a new issue here, since I think it's not SST related this time -> ...
